bokeh/core/has_props.py

```diff
@@ -607,8 +607,16 @@ def apply_theme(self, property_values):
 
         # Emit any change notifications that result
         for k, v in old_values.items():
-            prop = self.lookup(k)
-            prop.trigger_if_changed(self, v)
+            descriptor = self.lookup(k)
+            if isinstance(v, PropertyValueContainer) and v._unmodified_default_value:
+                if k in self._property_values:
+                    self._property_values[k]._unregister_owner(self, descriptor)
+                    del self._property_values[k]
+                if k in property_values:
+                    self._property_values[k] = descriptor.property.prepare_value(self.__class__, k, property_values[k])
+                    self._property_values[k]._unmodified_default_value = True
+                    self._property_values[k]._register_owner(self, descriptor)
+            descriptor.trigger_if_changed(self, v)
```

Comment:
OK this logic is simplified a bit, and it hopefully clearer. I could possibly live wth this (with more tests) but I'd still level to put things in a better place if there is one. The approach is now:

* if container prop value is in `self._property_values` and unmodified, it's only there because of the `getattr` above. So, "undo" that
* Then, if the container prop is in the new values "do" the business to put it in `self._property_values` so it show up as the new value if read from